### PR TITLE
Plugin/mix-extended

### PR DESCRIPTION
**Pull request by EricRovell**
Extending the functionality of `mix` plugin by generating mixtures of colors such as [tints, shades, and tones](https://en.wikipedia.org/wiki/Tints_and_shades) with `.tones(colors = 5)`, `.shades(colors = 5)`, and `.tones(colors = 5)` methods respectively.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -501,6 +501,54 @@ colord("#008080").mix("#808000", 0.35).toHex(); // "#50805d"
...
+
+lord("#ff0000").shades(10).map(color => color.toHex()) // ["#ff0000", "#e31004", "#c81707", "#ad1b09", "#931c0b", "#7a1b0b", "#621a0b", "#4a1709", "#341306", "#200d03", "#000000"];
+```
+
+</details>
+
+<details>
+  <summary><b><code>.shades(colors = 5)</code></b> (<b>mix</b> plugin)</summary>
+
+Provides functionality to generate [tones](https://en.wikipedia.org/wiki/Tints_and_shades) of a color. Returns an array of `Colord` instances, including the original color.
```
Comment (omgovich) [inline README.md:539]:
```suggestion

<details>
  <summary><b><code>.tones(colors = 5)</code></b> (<b>mix</b> plugin)</summary>

Provides functionality to generate [tones](https://en.wikipedia.org/wiki/Tints_and_shades) of a color. Returns an array of `Colord` instances, including the original color.
```

File: src/plugins/mix.ts [hunk 2]
```diff
@@ -21,6 +38,26 @@ const mixPlugin: Plugin = (ColordClass): void => {
...
+  ColordClass.prototype._mixPalette = function (color, colors = 5) {
+    const palette = [];
+    for (let ratio = 0; ratio <= 1; ratio += 1 / colors) {
+      palette.push(this.mix(color, ratio));
+    }
+    return palette;
+  };
+
+  ColordClass.prototype.tints = function (colors) {
+    return this._mixPalette("#ffffff", colors);
```
Comment (omgovich) [inline src/plugins/mix.ts:51]: To reduce a bundle size a bit:

```suggestion
    return this._mixPalette("#fff", colors);
```

File: src/plugins/mix.ts [hunk 3]
```diff
@@ -21,6 +38,26 @@ const mixPlugin: Plugin = (ColordClass): void => {
...
+  };
+
+  ColordClass.prototype.shades = function (colors) {
+    return this._mixPalette("#000000", colors);
```
Comment (omgovich) [inline src/plugins/mix.ts:55]: To reduce a bundle size a bit:
```suggestion
    return this._mixPalette("#000", colors);
```

File: README.md (hunk 1 above)
Comment (EricRovell) [reply README.md:539]: My bad :)

File: src/plugins/mix.ts [hunk 4]
```diff
@@ -21,6 +38,26 @@ const mixPlugin: Plugin = (ColordClass): void => {
...
+    const palette = [];
+    for (let ratio = 0; ratio <= 1; ratio += 1 / colors) {
+      palette.push(this.mix(color, ratio));
+    }
+    return palette;
+  };
+
+  ColordClass.prototype.tints = function (colors) {
+    return this._mixPalette("#ffffff", colors);
+  };
```
[the remark this answers is not in the document]
Comment (EricRovell) [reply src/plugins/mix.ts:52]: Yeah, it makes sense. Will do,

File: README.md [hunk 5]
```diff
@@ -827,6 +875,19 @@ colord("#cd853f").mix("#eee8aa", 0.6).toHex(); // "#e3c07e"
...
+Also, plugin provides special mixtures such as [tints, shades, and tones](https://en.wikipedia.org/wiki/Tints_and_shades):
+
+<div align="center">
+<image src="https://upload.wikimedia.org/wikipedia/commons/thumb/d/d6/Tint-tone-shade.svg/320px-Tint-tone-shade.svg.png" alt="tints, shades, and tones mixtures">
+</div>
+
+```js
+colord("#ff0000").tints(10).map(color => color.toHex()) // ["#ff0000", "#ff3f20", "#ff5c39", "#ff7551", "#ff8b68", "#ff9f80", "#ffb399", "#ffc6b2", "#ffd9cb", "#ffece5", "#ffffff"];
+colord("#ff0000").shades(10).map(color => color.toHex()) // ["#ff0000", "#e31004", "#c81707", "#ad1b09", "#931c0b", "#7a1b0b", "#621a0b", "#4a1709", "#341306", "#200d03", "#000000"];
+colord("#ff0000").tones(10).map(color => color.toHex()) // ["#ff0000", "#f52d14", "#ea4023", "#df4e30", "#d4583b", "#c86147", "#bc6952", "#af705e", "#a17669", "#917b75", "#808080"];
```
Comment (omgovich) [inline README.md:887]: I guess we can generate 4-5 colors here to keep the docs more readable. What do you think?

File: README.md [hunk 6]
```diff
@@ -809,7 +857,7 @@ colord("#646464").alpha(0.5).toLchString(); // "lch(42.37% 0 0 / 0.5)"
 <details>
   <summary><b><code>mix</code> (Color mixing)</b> <i>0.9 KB</i></summary>
```
Comment (omgovich) [inline README.md:858]: Could I ask you to run `npm run size` locally and update the size here? 🙏

File: README.md [hunk 7]
```diff
@@ -501,6 +501,54 @@ colord("#008080").mix("#808000", 0.35).toHex(); // "#50805d"
 
 </details>
 
+<details>
+  <summary><b><code>.tints(colors = 5)</code></b> (<b>mix</b> plugin)</summary>
```
Comment (omgovich) [inline README.md:505]: Let's maybe call the param count to make it clear that the user should pass a number here (not a color)?

File: src/plugins/mix.ts [hunk 8]
```diff
@@ -1,18 +1,34 @@
...
+
+    /**
+     * Generates a shades palette based on original color.
+     */
+    shades(colors: number): Colord[];
+
+    /**
+     * Generates a tones palette based on original color.
+     */
+    tones(colors: number): Colord[];
```
Comment (omgovich) [inline src/plugins/mix.ts:26]: Since it's optional.

```suggestion
    /**
     * Generates a tints palette based on original color.
     */
    tints(colors?: number): Colord[];

    /**
     * Generates a shades palette based on original color.
     */
    shades(colors?: number): Colord[];

    /**
     * Generates a tones palette based on original color.
     */
    tones(colors?: number): Colord[];
```

File: tests/plugins.test.ts [hunk 9]
```diff
@@ -275,6 +275,18 @@ describe("mix", () => {
...
+    expect(colord("#ff0000").tints(10).map(color => color.toHex())).toEqual(["#ff0000", "#ff3f20", "#ff5c39", "#ff7551", "#ff8b68", "#ff9f80", "#ffb399", "#ffc6b2", "#ffd9cb", "#ffece5", "#ffffff"]);
+  });
+
+  it("Generates a shades palette", () => {
+    expect(colord("#ff0000").shades(10).map(color => color.toHex())).toEqual(["#ff0000", "#e31004", "#c81707", "#ad1b09", "#931c0b", "#7a1b0b", "#621a0b", "#4a1709", "#341306", "#200d03", "#000000"]);
+  });
+
+  it("Generates a tones palette", () => {
+    expect(colord("#ff0000").tones(10).map(color => color.toHex())).toEqual(["#ff0000", "#f52d14", "#ea4023", "#df4e30", "#d4583b", "#c86147", "#bc6952", "#af705e", "#a17669", "#917b75", "#808080"]);
+  });
```
Comment (omgovich) [inline tests/plugins.test.ts:289]: Please add simple test cases to check that methods work properly with no arguments (e.g. `.tints()` should return 5 colors)

Comment (omgovich) [reply tests/plugins.test.ts:289]: And let's check edge-cases like 2 and 3 colors generation 🙏

File: README.md (hunk 5 above)
Comment (EricRovell) [reply README.md:887]: Sure, no need to have here so many colors :)

File: README.md (hunk 6 above)
Comment (EricRovell) [reply README.md:858]: Sure, will do!

**Review by EricRovell — COMMENTED** [review]
Got it. What should I do next to finish this?

**Comment by omgovich** [comment]
Nothing :) Going to merge this in a few days. Just didn't have time to double check it yet. Pretty busy with my new apartment :)

**Comment by EricRovell** [comment]
Ah, no problem at all. I just was thinking that there was some action required from me :)

**Comment by omgovich** [comment]
Great job! Thanks a lot for that! ❤️

**Comment by EricRovell** [comment]
Glad to help!
